src/midi/midi_mt32.cpp

```diff
@@ -683,101 +694,124 @@ void MidiHandler_mt32::Close()
 
 	// Reset the members
 	service.reset();
-	total_buffers_played = 0;
-	last_played_frame = 0;
 
-	is_open = false;
+	last_rendered_ms = 0.0;
+	ms_per_frame     = 0.0;
+
+	is_open          = false;
 }
 
-uint32_t MidiHandler_mt32::GetMidiEventTimestamp() const
+uint16_t MidiHandler_mt32::GetNumPendingFrames()
 {
-	const uint32_t played_frames = total_buffers_played * FRAMES_PER_BUFFER;
-	return service->convertOutputToSynthTimestamp(played_frames +
-	                                              last_played_frame);
+	const auto now_ms = PIC_FullIndex();
+
+	// Wake up the channel and update the last rendered time datum.
+	assert(channel);
+	if (channel->WakeUp()) {
+		last_rendered_ms = now_ms;
+		return 0;
+	}
+	if (last_rendered_ms >= now_ms) {
+		return 0;
+	}
+	// Return the number of frames needed to get current again
+	assert(ms_per_frame > 0.0);
+	const auto elapsed_ms = now_ms - last_rendered_ms;
+	const auto frames     = iround(ceil(elapsed_ms / ms_per_frame));
+	last_rendered_ms += (frames * ms_per_frame);
+	return check_cast<uint16_t>(frames);
 }
 
+// The request to play the channel message is placed in the MIDI work FIFO
 void MidiHandler_mt32::PlayMsg(const uint8_t *msg)
 {
-	assert(channel);
-	channel->WakeUp();
-
-	const auto msg_words = reinterpret_cast<const uint32_t *>(msg);
-	const std::lock_guard<std::mutex> lock(service_mutex);
-	service->playMsgAt(SDL_SwapLE32(*msg_words), GetMidiEventTimestamp());
+	constexpr auto channel_len = 4;
+	std::vector<uint8_t> message(msg, msg + channel_len);
```

Comment:
I'm confused by this syntax. Isn't the first parameter to the vector constructor the number of elements? But that's `msg`. Then we're adding the message and the length together, how does that work?

Reply:
Oh, this creates the vector in-place holding the C-based msg array. 

You construct with the first and last pointers (basically, begin() and end()).

The key here is to use `pointers to T` (for the start and end), and then the vector will accept it as bulk-data.

``` c++
vector<T> my_vec(const *T begin, const *T end);
```

Reply:
Right, that makes sense, thanks.